lib/combinat.gi

```diff
@@ -2498,18 +2495,130 @@ InstallGlobalFunction( IteratorOfPartitionsSet, function( s )
+          iter!.next := false;
+          return out;
+        fi;
+        # If the size of the partition equals sz we return, otherwise it is too small,
+        # and we continue to iterate until the next partition of the correct size is found
+        if Maximum(iter!.next)+1 = iter!.sz then
+          if m = iter!.sz then
+            return out;
+          else
+            part := StructuralCopy(iter!.next);
+          fi;
+        fi;
+      od;
+    end );
+
+BindGlobal( "NextIterator_PartitionsSetGivenSizeOrLess", function( iter )
+      local j, max, part, m, out, i;
+      if Length(iter!.s) = 0 or Length(iter!.s)<iter!.sz then
+        iter!.next := false;
+        return [];
+      fi;
+      part := StructuralCopy(iter!.next);
+      while true do
+        j := Size(iter!.next);
+        # Compute next restricted growth string using a_i \leq Max({a_0 .. a_{i-1}})+1
+        while j > 1 do
+          max := Minimum(Maximum(iter!.next{[1 .. j-1]}), iter!.sz-2);
+          if iter!.next[j] <= max then
+            iter!.next[j] := iter!.next[j]+1;
+            m := Maximum(iter!.next)+1;
+            break;
+          else
+            iter!.next[j] := 0;
+            j := j-1;
+          fi;
+        od;
+        m := Maximum(part)+1;
+        # Convert restricted growth string to partition of set
+        out := List([1 .. m], t -> []);
+        for i in [1 .. Size(part)] do
+          Add(out[part[i]+1], iter!.s[i]);
+        od;
+        # this is the final iteration if in the next iteration we cycle through to the start again
+        if ForAll([2 .. Size(iter!.next)], t -> iter!.next[t]=0) then
+          iter!.next := false;
+          return out;
+        fi;
+        return out;
+      od;
+    end );
 
-    shallowCopy := iter -> rec( next := ShallowCopy( iter!.next ), s := iter!.s );
+BindGlobal( "ShallowCopy_PartitionsSet",
+    iter -> rec( next := ShallowCopy( iter!.next ), s := iter!.s, sz := iter!.sz ) );
 
-    isDone := iter -> iter!.next = false;
+InstallGlobalFunction( IteratorOfPartitionsSet , function( s, arg... )
+    local k, r;
 
-    return IteratorByFunctions( rec(
-      IsDoneIterator := isDone,
-      NextIterator := nextIterator,
-      ShallowCopy := shallowCopy,
-      s := Immutable(s),
-      next := ListWithIdenticalEntries(Size(s), 0) ) );
+    if not IsSet(s) then
+      Error( "<s> must be a set" );
+    fi;
+
+    r := rec(
+            IsDoneIterator := IsDoneIterator_PartitionsSet,
+            NextIterator := NextIterator_PartitionsSet,
+            ShallowCopy := ShallowCopy_PartitionsSet,
+            s := Immutable(s),
+            next := ListWithIdenticalEntries(Size(s), 0),
+            sz := fail,
+         );
+
+    if Length( arg ) = 1 or Length( arg ) = 2 then
+      k := arg[1];
+      if not IsInt(k) then
+        Error("usage: <k> must be an integer");
+      elif k<1 or k > Length(s) then
+        Error("usage: <k> must be between 1 and size of <s>");
```

Comment:
Due to time constraints, I would prefer not to update PartitionsSet to support flag myself at the moment. I think it would be best to submit an issue.

Reply:
I have modified the iterator so that it matches PartitionsSet when k is less than 1 or greater than the size of the set. I replaced the error message with the following:

        if s <> [] or k <> 0 then
          r.next:=false;
        fi;

the case where s=[] and k=0, is the only time [] is returned. Otherwise nothing ought to be returned, and so setting r.next=false means that calling IsDoneIterator will immediately return true. If you were now to use a while loop to iterate through the examples you gave, the iterator would produce the same output.

Reply:
The same thing should be done when k is not an integer. I just updated the code in the same way.

Reply:
Actually it would have been fine to leave in the check for `IsInt(k)` and the subsequent error: yes, `PartitionsSet` silently accepts those, but only because it simply fails to check for this case at all. Overall, I think it'd be better if both functions gave an error if `k` isn't an integer (I am sure all those cases are user mistakes and a helpful error may help prevents lots of wasted time searching for a bug). That said, I am not asking (and much less requiring!) you to do either (you already went way up and above all expectations on this PR, thanks for that), just wanted to mention it: there is no need for "bug by bug compatibility" here ;-)

Reply:
Should I change it back then? I'm happy to do that here. I might leave the PartititionsSet error message for someone else to implement though